### PR TITLE
Entity Implementation

### PR DESCRIPTION
**Pull request by ruihan00**
Rough implementation of entity
- Added Components
- Added PlayerEntity
- Added EntityManager

Things to consider
- entity may be either sprite sheet or single image, either sprite component will handle that or we have 2 components 
- some way to represent sprite sheet and animation in sprite component

### CODE REVIEW

File: star-dash/star-dash/Entities/Components/PositionComponent.swift [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+//
+//  Created by Lau Rui han on 12/3/24.
+//
+
+import Foundation
+
+
+class PositionComponent: Component {
+    var position: CGPoint
+    var rotation: Float
```
Comment (jasonqiu212) [inline star-dash/star-dash/Entities/Components/PositionComponent.swift:13]: should rotation fall under a separate component on its own?

File: star-dash/star-dash/Entities/Components/PositionComponent.swift [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        self.init(id: UUID(), entityId: entityId, position: position, rotation: rotation)
+    }
+    
+    func setPosition(position: CGPoint) {
+        self.position = position
+    }
+    
+    func setRotation(rotation: Float) {
+        self.rotation = rotation
+    }
```
Comment (jasonqiu212) [inline star-dash/star-dash/Entities/Components/PositionComponent.swift:31]: do we still need these setter methods, if we are leaving the `position` and `rotation` attributes as public?

File: star-dash/star-dash/Entities/EntityManager.swift [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+//  Created by Lau Rui han on 12/3/24.
+//
+
+import Foundation
+typealias ComponentSet = Set<ComponentId>
+class EntityManager {
+    var componentMap: [ComponentId: Component]
+    var entityMap: [EntityId: Entity]
+    var entityComponentMap: [EntityId: ComponentSet]
+    init(componentMap: [ComponentId : Component], entityMap: [EntityId : Entity], entityComponentMap: [EntityId: ComponentSet]) {
```
Comment (jasonqiu212) [inline star-dash/star-dash/Entities/EntityManager.swift:14]: I think can set up typealising for these complex types for improved readability

```
typealias ComponentMap = [ComponentId: Component]
typealias EntityMap = [EntityId: Entity]
typealias EntityComponentsMap = [EntityId: ComponentSet]
```

File: star-dash/star-dash/Entities/Components/SpriteComponent.swift [hunk 4]
```diff
@@ -0,0 +1,24 @@
+//
+//  SpriteComponent.swift
+//  star-dash
+//
+//  Created by Lau Rui han on 12/3/24.
+//
+
+import Foundation
+class SpriteComponent: Component {
+    // sprite could be single image or sprite set, could use enum to seperate, for sprite set will need to discuss how to rep animation
```
[the remark this answers is not in the document]
Comment (ruihan00) [reply star-dash/star-dash/Entities/Components/SpriteComponent.swift:10]: maybe this one wait until junyi start on the sprites then we can discuss the representation, im not sure how it needs to be represented

File: star-dash/star-dash/Entities/Components/Component.swift [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+//  star-dash
+//
+//  Created by Lau Rui han on 12/3/24.
+//
+
+import Foundation
+typealias ComponentId = UUID
+class Component {
+    var entityId: UUID
+    var id: UUID
```
[the remark this answers is not in the document]
Comment (ruihan00) [reply star-dash/star-dash/Entities/Components/Component.swift:12]: now sure if storing direct reference to entity is needed, storing id is more memory efficient

File: star-dash/star-dash/Entities/EntityManager.swift (hunk 3 above)
Comment (ruihan00) [reply star-dash/star-dash/Entities/EntityManager.swift:14]: changed

File: star-dash/star-dash/Entities/Entity.swift [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+//  star-dash
+//
+//  Created by Lau Rui han on 12/3/24.
+//
+
+import Foundation
+typealias EntityId = UUID
+
+protocol Entity {
+    var id: UUID {get}
```
[the remark this answers is not in the document]
Comment (ruihan00) [reply star-dash/star-dash/Entities/Entity.swift:12]: changed

**Comment by ruihan00** [comment]
added entity implementation for the rest of the entities, missing test for now

File: star-dash/star-dash/Entities/Components/PositionComponent.swift (hunk 1 above)
Comment (jasonqiu212) [reply star-dash/star-dash/Entities/Components/PositionComponent.swift:12]: Can keep for now. If the need arises in the future, can separate later

**Review by jasonqiu212 — APPROVED** [review]
LGTM!

**Pull request merged**